### PR TITLE
Add links for browsing to a specified directory

### PR DESCRIPTION
**Pull request by dbenage-cx**
Adds a text tag that supports opening a users preferred file manager at a configured directory.
(Usage: `<browsepath `*type*`>`*optional appended text*`</browsepath>`)

Initial purpose of this PR is to support a clickable link for a users config directory within a pedia article for #1552 

Testing requested for verification on Windows and OSX.
e.g. within the stringtable entry for a pedia article or sitrep add any or all of:
```
<browsepath PATH_CONFIG>Config Files</browsepath>
<browsepath PATH_DATA_USER/shipdesigns>Custom Ship Designs</browsepath>
<browsepath PATH_DATA_ROOT>Browse:PATH_DATA_ROOT</browsepath>
<browsepath PATH_RESOURCE></browsepath>
<browsepath PATH_RESOURCE/python/AI>PATH_RESOURCE/python/AI</browsepath>
[[browsepath PATH_SAVE]]
```
Clicking on the link in-game should open the preferred file manager at your configured user data directory.

### CODE REVIEW

**Comment by dbenage-cx** [comment]
MSVC will likely fail to link due to needing project file update.
(parallels no longer functional, looking at alternatives)

**Comment by dbenage-cx** [comment]
Attempting `open` for windows command and reverting to `system()`
I did not expect pulling in window.h for `ShellExecute` (and saw some posts citing odd behaviour in win7 & 8)

**Comment by geoffthemedio** [comment]
````
C:\Users\Geoff\Desktop\FreeOrionSDK5\FreeOrion>FreeOrion
'open' is not recognized as an internal or external command,
operable program or batch file.
````

**Comment by dbenage-cx** [comment]
@geoffthemedio TY for testing
Trying `start` with an appended path separator
(Otherwise given dir `X` with a sub-dir `Y` and a file `Y.COM`: `start X/Y` may execute `Y.COM`)

**Comment by geoffthemedio** [comment]
````
DIPLOMACY_TEXT
'''
<browsepathtype> optional appended text</browsepath>

Non-exhaustive list of Diplomacy features:
'''
````
gives me

````
Error parsing rich text tags: expected end tag:browsepathtype got: "

Non-e..."
````

````
DIPLOMACY_TEXT
'''
<browsepath> optional appended text</browsepath>

Non-exhaustive list of Diplomacy features:
'''
````
gives me a clickable link with text `ERROR`. Clicking a link opens a directory browse window to `C:\Users\Geoff\AppData\Local\Temp` which is not `C:\Users\Geoff\AppData\Roaming\FreeOrion` which is where config.xml is.

Changing it to
````
DIPLOMACY_TEXT
'''
<browsepath %APPDATA%/FreeOrion> optional appended text</browsepath>

Non-exhaustive list of Diplomacy features:
'''
````
Still produces an `ERROR` link that takes me to `C:\Users\Geoff\AppData\Local\Temp`

**Comment by dbenage-cx** [comment]
@geoffthemedio *type* is a string representation of enum PathType.
If given an invalid PathType, the link points to the (os dependant) temporary directory and displayed string is `ERROR`.
e.g. `<browsepath PATH_BINARY></browsepath>`
I've added a commit to support the form `[[browsepath `*`PathType`*`]]` (trailing space is md format error)

The current PathTypes are:
```
    PATH_BINARY
    PATH_RESOURCE
    PATH_PYTHON
    PATH_DATA_ROOT
    PATH_DATA_USER
    PATH_CONFIG
    PATH_SAVE
    PATH_TEMP
```

**Comment by geoffthemedio** [comment]
I added links to all the listed path types to the "Researching Technologies" article stringtable entry, then set my save path to `C:\Users\Geoff\Desktop\FreeOrionSDK5\FreeOrion\asdfШит`

In the pedia, this appears as `
C:\Users\Geoff\Desktop\FreeOrionSDK5\FreeOrion\asdf???\`

You probably need to use the PathString function, and not assume paths are stored as and convert simply to UTF-8.

**Comment by geoffthemedio** [comment]
While not essential for this pull request, the path browse linking should also support absolute and composed paths. For example, I should be able to put links in the Text - Textures article to the individual file names, by doing something like

`[[browsepath C:/Users/Geoff/Desktop/FreeOrionSDK5/FreeOrion/default/data/art/icons/buttons/charts.png]]`

I should also be able to link to a directory within one of the built-in paths, by doing something like

`[[browsepath PATH_RESOURCE/customizations]]'

**Comment by dbenage-cx** [comment]
Rebased.
Using `PathString` for any conversions.

I did not make further modifications for stringtable lookup (`[[browsepath PATH_SAVE/subdir]]`), however the form `<browsepath PATH_SAVE/subdir></browsepath>` is now available.
An empty label text will be populated with the path.

Also modifying any matches within the label text area
`<browsepath PATH_SAVE/subdir>`*`optional_text`*`PATH_SAVE`*`optional_text`*`</browsepath>`.
Outside of an empty entry or containing a PathType, the label text would be unmodified.

The form `<browsepath some_raw_dir>`*`label`*`</browsepath>` does not fill in any empty label.

When using a custom subdir or raw dir, if the directory(or parent directory if a file) does not exist, BrowsePath will early return with an error logged.

tl;dr some valid forms:
* `[[browsepath PATH_TYPE]]`
* `<browsepath PATH_TYPE></browsepath>`
* `<browsepath PATH_TYPE>label</browsepath>`
* `<browsepath PATH_TYPE>pre-text PATH_TYPE post-text</browsepath>`
* `<browsepath PATH_TYPE/subdir>label</browsepath>`
* `<browsepath PATH_TYPE/subdir>PATH_TYPE/subdir</browsepath>`
* `<browsepath rawdir>label</browsepath>`

**Comment by geoffthemedio** [comment]
Giving usable examples would be a lot more helpful than a formula for making them... Regardless, I added
````
    [[browsepath PATH_DATA_ROOT]]
    <browsepath PATH_DATA_ROOT></browsepath>
    <browsepath PATH_DATA_ROOT>label</browsepath>
    <browsepath PATH_DATA_ROOT>pre-text PATH_DATA_ROOT post-text</browsepath>
    <browsepath PATH_DATA_ROOT/data/art>label</browsepath>
    <browsepath PATH_DATA_ROOT/data/art>PATH_DATA_ROOT/data/art</browsepath>
    <browsepath C:\Users>Users</browsepath>
    <browsepath C:\users>users</browsepath>
````
to an article. The result is a few blank lines, then 6 lines of text coloured like link but not clickable. (No mouseoverhighlight, no response to clicking)

links

**Comment by dbenage-cx** [comment]
Reworked to separate multiple links in text, rebased and updated description with some exact examples.

**Comment by geoffthemedio** [comment]
Same test case as above. I now get:
````
[[browsepath PATH_DATA_ROOT]]

label
pre-text PATH_DATA_ROOT post-text
label
PATH_DATA_ROOT/data/art
Users
users
````
The lines after `label` are coloured like links, but nothing is clickable.

**Comment by dbenage-cx** [comment]
TY for testing, I will try to track down a windows box to troubleshoot.

**Comment by geoffthemedio** [comment]
@dbenage-cx bump / reminder

**Comment by dbenage-cx** [comment]
Apologies for long delay, on reviewing this was not a windows issue, rather pedia articles did not resolve correctly (I must have only been checking sitreps).

The following example diff shows changes in initial turn sitreps and pedia(Guides->Greetings):
```diff
diff --git a/default/stringtables/en.txt b/default/stringtables/en.txt
index 270c2fe80..9b1e08a6a 100644
--- a/default/stringtables/en.txt
+++ b/default/stringtables/en.txt
@@ -5300,7 +5300,14 @@ GREETINGS_INTRO
 Greetings Esteemed Entity, and Welcome to the FreeOrion Universe!
 
 GREETINGS_GUIDE_REFERENCE
-Newcomers to this universe may particularly wish to consult the [[encyclopedia QUICK_START_GUIDE_TITLE]].
+'''Newcomers to this universe may particularly wish to consult the [[encyclopedia QUICK_START_GUIDE_TITLE]].
+<browsepath PATH_CONFIG>Config Files</browsepath>
+<browsepath PATH_DATA_USER/shipdesigns>Custom Ship Designs</browsepath>
+<browsepath PATH_DATA_ROOT>Browse: PATH_DATA_ROOT</browsepath>
+<browsepath PATH_RESOURCE></browsepath>
+<browsepath PATH_RESOURCE/python/AI>PATH_RESOURCE/python/AI</browsepath>
+[[browsepath PATH_SAVE]]
+'''
 
 
 GREETINGS_GUIDE_TEXT
```

Rebased with changes.

**Comment by geoffthemedio** [comment]
I added to a pedia article:
````

<browsepath PATH_CONFIG>Config Files</browsepath>
<browsepath PATH_DATA_USER/shipdesigns>Custom Ship Designs</browsepath>
<browsepath PATH_DATA_ROOT>Browse: PATH_DATA_ROOT</browsepath>
<browsepath PATH_RESOURCE></browsepath>
<browsepath PATH_RESOURCE/python/AI>PATH_RESOURCE/python/AI</browsepath>
[[browsepath PATH_SAVE]]

    [[browsepath PATH_DATA_ROOT]]
    <browsepath PATH_DATA_ROOT></browsepath>
    <browsepath PATH_DATA_ROOT>label</browsepath>
    <browsepath PATH_DATA_ROOT>pre-text PATH_DATA_ROOT post-text</browsepath>
    <browsepath PATH_DATA_ROOT/data/art>label</browsepath>
    <browsepath PATH_DATA_ROOT/data/art>PATH_DATA_ROOT/data/art</browsepath>
    <browsepath C:\Users>Users</browsepath>
    <browsepath C:\users>users</browsepath>
    
````
In game, it looked reasonable, and most links worked:
````

Config Files
Custom Ship Designs
Browse: C:\Users\Geoff\Desktop\FreeOrionSDK5\FreeOrion
C:\Users\Geoff\Desktop\FreeOrionSDK5\FreeOrion\default
C:\Users\Geoff\Desktop\FreeOrionSDK5\FreeOrion\default/python/AI
C:\Users\Geoff\AppData\Roaming\FreeOrion\save

    C:\Users\Geoff\Desktop\FreeOrionSDK5\FreeOrion
    C:\Users\Geoff\Desktop\FreeOrionSDK5\FreeOrion
    label
    pre-text C:\Users\Geoff\Desktop\FreeOrionSDK5\FreeOrion post-text
    label
    C:\Users\Geoff\Desktop\FreeOrionSDK5\FreeOrion/data/art
    Users
    users
````
Links that didn't work were the `Custom Ship Designs`, and the last `label` and the  `art` link after it. That mostly makes sense since those directories don't exist; I have no custom ship directory, and the latter two are lacking `default/` in the path. After creating and saving a custom ship design, that link works.

That said, some way to handle broken links might be useful?

**Comment by geoffthemedio** [comment]
I set my save directory to `C:\Users\Geoff\Desktop\testД` which exists (`C:\Users\Geoff\Desktop\testД` just copied from the folder window path bar).

When I click the link to that dir in the pedia article, it does nothing. In the log, I get:
`22:37:37.909533 [error] client : HumanClientApp.cpp:1327 : Can not browse to non-existant path: C:\Users\Geoff\Desktop\testД`

If I remove the Д from the path and option setting, it works. I suspect your code doesn't handle window path format conversions properly.

**Comment by dbenage-cx** [comment]
Thank you for catching the conversions, hopefully caught such cases.
Added a confirmation dialog to attempt a parent directory, if target directory does not exist.

**Comment by geoffthemedio** [comment]
Confirmation dialog not working correctly for me...

blank_error

Don't think a dialog is necessary though... just default to the parent window?

**Comment by geoffthemedio** [comment]
I don't think a dialog is needed... it's reasonable to just open the next existing container directory in the specified path.

**Comment by geoffthemedio** [comment]
Now when I click on a bad directory link, I get an instant crash. In the log there is a bad weak pointer exception:
````chmain.cpp:275 : main() caught exception(std::exception): bad_weak_ptr````

Stepping through execution, the crash happens here:
https://github.com/dbenage-cx/freeorion/blob/1a72d2345c7a9d9d16c19181c0b3a6fc44ba2d35/client/human/HumanClientApp.cpp#L1336

**Comment by dbenage-cx** [comment]
Ok, replaced dialog with recursive call for parent path.

File: UI/LinkText.h [hunk 1]
```diff
@@ -172,6 +180,7 @@ class LinkText : public GG::TextControl, public TextLinker {
 
 private:
     void SetLinkedText(const std::string& str) override;
+    //void Init(bool resize = false);
```
Comment (geoffthemedio) [inline UI/LinkText.h]: can this be removed?

Comment (dbenage-cx) [reply UI/LinkText.h]: Fixed

**Comment by geoffthemedio** [comment]
@Vezzra could you test on OSX?

File: client/human/HumanClientApp.cpp [hunk 2]
```diff
@@ -1316,3 +1316,57 @@ void HumanClientApp::OpenURL(const std::string& url) {
...
+    if (full_path.empty()) {
+        ErrorLogger() << "Unable to determine directory for path " << browse_path.string();
+        return;
+    }
+
+// Append trailing slash to prevent executing a file with same name(minus extension) as folder
+#if defined(_WIN32)
+    system(("start /b " + PathString(full_path) + "\\").c_str());
+#elif defined(__APPLE__)
+    system(("open " + PathString(full_path) + "/ &").c_str());
```
Comment (Vezzra) [inline client/human/HumanClientApp.cpp]: On OSX this does not work when the path contains white spaces. You need to wrap the path in double quotes:
```
    system(("open \"" + PathString(full_path) + "/\" &").c_str());
```
I'd expect the same problem on the other platforms, but didn't test...

**Review by Vezzra — CHANGES_REQUESTED** [review]
Two issues on OSX:

* Paths containing white spaces don't work (see my other comment with suggested fix).

* Custom paths (like `/foo/bar`) don't work: If you don't specify a label, nothing is displayed at all, if you do, the label is printed, but when you mouse over it, it doesn't get highlighted, and when you click on it, nothing happens.

Everything else seems to work as intended.

File: client/human/HumanClientApp.cpp (hunk 2 above)
Comment (dbenage-cx) [reply client/human/HumanClientApp.cpp]: TY, added quotes to path target.

**Comment by dbenage-cx** [comment]
@Vezzra TY for testing.  Currently a raw path requires a label, (not certain what use case these have, as they will not be valid on at least one platform).  Do you recall the entry you added that displayed a bad/non-linking label?

**Comment by Vezzra** [comment]
>  Do you recall the entry you added that displayed a bad/non-linking label?

IIRC, that has been `/Users/user`, which is my home directory on my Mac.

As you correctly pointed out, the usefulness of raw paths is highly questionable, because they cannot work platform-independently. However, if you provide the feature, it should work, otherwise it's probably better to remove it.

**Comment by geoffthemedio** [comment]
So is this working and ready for merging now? @Vezzra ?

**Comment by Vezzra** [comment]
Everything but raw paths is working now (on OSX). The raw path (or better, the specified label) gets highlighted now when moused-over, but clicking on it still does nothing.

**Comment by dbenage-cx** [comment]
With an invalid param, such as raw path, the label should now be prefixed with (translated) `ERROR:`, with no action on click.
Previously an invalid param broke later links within the same text, which should now also be resolved.

**Comment by geoffthemedio** [comment]
Did almost the same test as posted above. I get most of the the links in the pedia, and mousing over highlights, but clicking on any does nothing.

The absolute path links appear like `ERROR: Users`.

The log has a bunch of errors about invalid paths.
````
17:27:36.395629 [error] client : LinkText.cpp:427 : Invalid param "C:\Users" for tag browsepath
17:27:36.395629 [error] client : LinkText.cpp:427 : Invalid param "C:\Users\Geoff\Desktop\FreeOrionSDK5\asdfШит" for tag browsepath
17:27:36.404649 [error] client : LinkText.cpp:427 : Invalid param "C:\Users" for tag browsepath
17:27:36.405635 [error] client : LinkText.cpp:427 : Invalid param "C:\Users\Geoff\Desktop\FreeOrionSDK5\asdfШит" for tag browsepath
17:27:36.409653 [error] client : LinkText.cpp:427 : Invalid param "C:\Users" for tag browsepath
17:27:36.409653 [error] client : LinkText.cpp:427 : Invalid param "C:\Users\Geoff\Desktop\FreeOrionSDK5\asdfШит" for tag browsepath
17:27:36.413654 [error] client : LinkText.cpp:427 : Invalid param "C:\Users" for tag browsepath
17:27:36.413654 [error] client : LinkText.cpp:427 : Invalid param "C:\Users\Geoff\Desktop\FreeOrionSDK5\asdfШит" for tag browsepath
17:27:36.416656 [error] client : LinkText.cpp:427 : Invalid param "C:\Users" for tag browsepath
17:27:36.417656 [error] client : LinkText.cpp:427 : Invalid param "C:\Users\Geoff\Desktop\FreeOrionSDK5\asdfШит" for tag browsepath
17:27:36.420658 [error] client : LinkText.cpp:427 : Invalid param "C:\Users" for tag browsepath
17:27:36.420658 [error] client : LinkText.cpp:427 : Invalid param "C:\Users\Geoff\Desktop\FreeOrionSDK5\asdfШит" for tag browsepath
17:27:36.423660 [error] client : LinkText.cpp:427 : Invalid param "C:\Users" for tag browsepath
17:27:36.424661 [error] client : LinkText.cpp:427 : Invalid param "C:\Users\Geoff\Desktop\FreeOrionSDK5\asdfШит" for tag browsepath
17:27:40.907991 [debug] client : HumanClientApp.cpp:1331 : Non-existant path: PATH_INVALID trying parent directory
17:27:42.690582 [debug] client : HumanClientApp.cpp:1331 : Non-existant path: PATH_INVALID trying parent directory
17:27:43.685120 [debug] client : HumanClientApp.cpp:1331 : Non-existant path: PATH_INVALID trying parent directory
17:27:44.880134 [debug] client : HumanClientApp.cpp:1331 : Non-existant path: PATH_INVALID trying parent directory
17:27:45.142376 [debug] client : HumanClientApp.cpp:1331 : Non-existant path: PATH_INVALID trying parent directory
17:27:45.882495 [debug] client : HumanClientApp.cpp:1331 : Non-existant path: PATH_INVALID trying parent directory
17:27:46.076389 [debug] client : HumanClientApp.cpp:1331 : Non-existant path: PATH_INVALID trying parent directory
17:27:46.534046 [debug] client : HumanClientApp.cpp:1331 : Non-existant path: PATH_INVALID trying parent directory
17:27:46.738033 [debug] client : HumanClientApp.cpp:1331 : Non-existant path: PATH_INVALID trying parent directory
17:27:47.187171 [debug] client : HumanClientApp.cpp:1331 : Non-existant path: PATH_INVALID trying parent directory
17:27:47.379185 [debug] client : HumanClientApp.cpp:1331 : Non-existant path: PATH_INVALID trying parent directory
````

I'm not sure how I'm supposed to specify the absolute paths, but with and without quotes didn't work. `C:\Users\Geoff\Desktop\FreeOrionSDK5\asdfШит` does exist.

After removing the absolute path links, I still can't click links, and get in the log:
````
17:35:27.009687 [debug] client : HumanClientApp.cpp:1331 : Non-existant path: C:\Users\Geoff\Desktop\FreeOrionSDK5\FreeOrionAlt/data/art trying parent directory
17:35:27.009687 [debug] client : HumanClientApp.cpp:1331 : Non-existant path: C:\Users\Geoff\Desktop\FreeOrionSDK5\FreeOrionAlt/data trying parent directory
````

**Comment by dbenage-cx** [comment]
Rebased to master, there is still an unresolved issue with links not highlighting in pedia entries, so marking as WIP until I can figure out why.

For absolute paths:  I had added extra logic to treat absolute paths as an error per previous comments.  I can revert that, but do not see a clear resolution to not requiring a label for them.

> Non-existant path: C:\Users\Geoff\Desktop\FreeOrionSDK5\FreeOrionAlt/data/art trying parent directory

I've changed the log statement to use PathString and note when aborting rather than trying a parent path.  It still should have opened the file manager, even if only at `C:\`
Did you mean to specify `PATH_RESOURCE` here, but used `PATH_DATA_DIR` ?
If `FreeOrionAlt` is a root data dir, it does not contain a `data` sub-directory (rather `default/data`)

Newest commit is to catch cases of spaces within a relative specified directory.
Provided user data directory contains an existing sub-directory `new dir name`:
`<browsepath PATH_DATA_USER/new dir name>directory with spaces</browsepath>`

**Comment by geoffthemedio** [comment]
Some of the test cases are supposed to be invalid paths, to test what happens in that case.

**Comment by geoffthemedio** [comment]
Now when I click a link, it seems to run `cmd.exe` again in the same window, but doesn't open a folder browse window.

The mouseover highlighting isn't working.

It doesn't show the `asdfШит` part of the path `<browsepath PATH_SAVE/asdfШит></browsepath>` in the pedia window. I just see `C:\Users\Geoff\AppData\Roaming\FreeOrion\save`

Test case was the following added to the end of the system blockade article:
````

<browsepath PATH_CONFIG>Config Files</browsepath>
<browsepath PATH_DATA_USER/shipdesigns>Custom Ship Designs</browsepath>
<browsepath PATH_DATA_ROOT>Browse: PATH_DATA_ROOT</browsepath>
<browsepath PATH_RESOURCE></browsepath>
<browsepath PATH_RESOURCE/python/AI>PATH_RESOURCE/python/AI</browsepath>
[[browsepath PATH_SAVE]]

<browsepath PATH_SAVE/asdfШит></browsepath>

    [[browsepath PATH_DATA_ROOT]]
    <browsepath PATH_DATA_ROOT></browsepath>
    <browsepath PATH_DATA_ROOT>label</browsepath>
    <browsepath PATH_DATA_ROOT>pre-text PATH_DATA_ROOT post-text</browsepath>
    <browsepath PATH_DATA_ROOT/data/art>label</browsepath>
    <browsepath PATH_DATA_ROOT/data/art>PATH_DATA_ROOT/data/art</browsepath>
    <browsepath C:\Users>Users</browsepath>
    <browsepath C:\users>users</browsepath>

````

`C:\Users\Geoff\AppData\Roaming\FreeOrion\save\asdfШит` exists.

**Comment by MatGB** [comment]
> there is still an unresolved issue with links not highlighting in pedia entries, so marking as WIP until I can figure out why

This appears to be a problem in master itself so is probably unrelated, neither link highlighting nor owned colour changing is working for me at the moment; I noticed both at the same time so suspect they're related but might be completely independent.

**Comment by dbenage-cx** [comment]
@MatGB Thank you, created an issue.

@geoffthemedio Appreciate your continued testing.  I've reworked the decoration to use xpressive regex.
If/when you find time again, a couple of additional checks might be a path containing at least one space, and an absolute path without a label.
E.G. (The later should get a label prefixed with `ERROR: `, same as with a label):
```
<browsepath PATH_SAVE/old saves>PATH_SAVE/old saves</browsepath>
<browsepath C:\users></browsepath>
```

Rebased

**Comment by geoffthemedio** [comment]
The `<browsepath PATH_SAVE/asdfШит></browsepath>` test case now shows up seemingly correctly in the pedia:
`C:\Users\Geoff\AppData\Roaming\FreeOrion\save/asdfШит`

Clicking any of the links just runs `cmd.exe` within the existing terminal window, and doesn't open the requested folder as correctly occurred in previous versions.

Your two new test cases show up as:
````
C:\Users\Geoff\AppData\Roaming\FreeOrion\save/old saves
ERROR: C:\users
````

**Comment by dbenage-cx** [comment]
Not fully reverted to previous code for `system` arguments, but a little more inline with those in `OpenURL`.

**Comment by geoffthemedio** [comment]
Clicking links still doesn't work... seems to just run another cmd.exe in the existing terminal window.

Edit: Might have been a git issue... rebuilding...
Edit2: Nope, still broken

**Comment by dbenage-cx** [comment]
Not certain where the failure is here, I'd be curious to know the log output for `Sending OS request to browse directory:` as it the result sounds akin to an empty argument for system.
If that line is correctly formed, the remainder of that line should have the desired effect if typed in the terminal window.

**Comment by geoffthemedio** [comment]
Now getting some but not all links clickable...
pedia_links

After clicking a link, I have this at the end of the log:
````
11:30:48.937597 [debug] client : ScopedTimer.cpp:168 : EncyclopediaDetailPanel::DoLayout - description->SizeMove time:       24 ms
11:30:48.937597 [debug] client : ScopedTimer.cpp:178 : EncyclopediaDetailPanel::DoLayout                         time:       26 ms
11:30:48.944619 [error] client : LinkText.cpp:404 : Invalid param "C:\Users" for tag browsepath
11:30:48.944619 [error] client : LinkText.cpp:404 : Invalid param "C:\users" for tag browsepath
11:30:48.950619 [error] client : LinkText.cpp:404 : Invalid param "C:\Users" for tag browsepath
11:30:48.950619 [error] client : LinkText.cpp:404 : Invalid param "C:\users" for tag browsepath
11:30:48.956610 [error] client : LinkText.cpp:404 : Invalid param "C:\Users" for tag browsepath
11:30:48.957611 [error] client : LinkText.cpp:404 : Invalid param "C:\users" for tag browsepath
11:30:48.963628 [error] client : LinkText.cpp:404 : Invalid param "C:\Users" for tag browsepath
11:30:48.963628 [error] client : LinkText.cpp:404 : Invalid param "C:\users" for tag browsepath
11:30:48.966629 [debug] client : ScopedTimer.cpp:168 : EncyclopediaDetailPanel::DoLayout - description->SizeMove time:       25 ms
11:30:48.966629 [debug] client : ScopedTimer.cpp:178 : EncyclopediaDetailPanel::DoLayout                         time:       25 ms
11:30:52.338942 [info] client : HumanClientApp.cpp:1542 : Sending OS request to browse directory: start /b "C:/Users\Geoff\AppData\Roaming\FreeOrion\save\asdfШит\"
11:30:59.902324 [debug] client : HumanClientApp.cpp:1037 : HumanClientApp::HandleFocusChange(Lost Focus)
````

**Comment by geoffthemedio** [comment]
Errors from AppVeyor reproduce for me locally.

**Comment by geoffthemedio** [comment]
Builds OK. Clicking links still doesn't work.
````
11:12:10.049725 [debug] client : HumanClientApp.cpp:1043 : HumanClientApp::HandleFocusChange(Gained Focus)
11:12:17.753157 [info] client : HumanClientApp.cpp:1529 : Sending OS request to browse directory: C:\Users\Geoff\Desktop\FreeOrionSDK5\FreeOrionAlt (C:\Users\Geoff\Desktop\FreeOrionSDK5\FreeOrionAlt)
11:12:19.680471 [debug] client : HumanClientApp.cpp:1043 : HumanClientApp::HandleFocusChange(Lost Focus)
````

**Comment by dbenage-cx** [comment]
Thanks for continuing to test, I'll try just `start path` if still not working.

**Comment by geoffthemedio** [comment]
No apparent change...
````
14:05:45.327483 [error] client : LinkText.cpp:404 : Invalid param "C:\users" for tag browsepath
14:05:48.920933 [info] client : HumanClientApp.cpp:1543 : Sending OS request to browse directory: start /b "C:\Users\Geoff\AppData\Roaming\FreeOrion\save\asdfШит\"
14:05:48.928939 [debug] client : HumanClientApp.cpp:1549 : std::system(start /b "C:\Users\Geoff\AppData\Roaming\FreeOrion\save\asdfШит\") returned 0
14:05:52.653043 [debug] client : HumanClientApp.cpp:1043 : HumanClientApp::HandleFocusChange(Lost Focus)
````

**Comment by geoffthemedio** [comment]
Something different! Clicking a test link gave me a popup from the console window:
cant_find

File: client/human/HumanClientApp.cpp [hunk 3]
```diff
@@ -1479,3 +1480,75 @@ void HumanClientApp::OpenURL(const std::string& url) {
...
+        return;
+    }
+
+    if (full_path.empty()) {
+        ErrorLogger() << "Unable to determine directory for path " << PathToString(full_path);
+        return;
+    }
+
+    full_path.make_preferred();
+    std::string target(PathToString(full_path));
```
Comment (geoffthemedio) [inline client/human/HumanClientApp.cpp]: This returns a UTF-8 string. For calling the Win32 API functions, you actually need a UTF-16 wstring. Possibly `fs::path::string_type native_string = full_path.native();`

Comment (geoffthemedio) [reply client/human/HumanClientApp.cpp]: Works for me:
````
    full_path.make_preferred();
    auto target = full_path.native();
    decltype(target) command;

    // Ampersand post-fixed to non-Windows platforms, prevents blocking until process exits
    // Double quotes around target to support paths containing spaces
    // Trailing slash post-fixed to prevent executing a file with same name(minus extension) as folder
#ifdef _WIN32
    command = L"start " + target;
#elif __APPLE__
    command = "open \"" + target + "/\" &";
#else
    command = "xdg-open \"" + target + "/\" &";
#endif

#ifdef _WIN32
    // https://msdn.microsoft.com/en-us/library/277bwbdz.aspx remarks all streams must be flushed or closed prior to std::system
    std::fflush(NULL);

    std::string utf8_command;
    utf8::utf16to8(command.begin(), command.end(), std::back_inserter(utf8_command));
    InfoLogger() << "Sending OS request to browse directory: " << utf8_command;

    int syscall_retval = _wsystem(command.c_str());

    DebugLogger() << "std::system(" << utf8_command << ") returned " << syscall_retval;

#else
    InfoLogger() << "Sending OS request to browse directory: " << command;
    int syscall_retval = std::system(command.c_str());
    DebugLogger() << "std::system(" << command << ") returned " << syscall_retval;
#endif
}
````

Comment (dbenage-cx) [reply client/human/HumanClientApp.cpp]: Oh progress! I'd like to try codecvt over utf8 lib.
Conceptually, the path could be imbued and allow `path.string()` instead of `PathToString(path)`.  Since I'm not able to test that, I've only used it for the one log conversion.
Failing this one attempt, will use a more verbose form of your changes.

**Comment by dbenage-cx** [comment]
Rebased

**Comment by geoffthemedio** [comment]
Like before, doesn't work.
`09:34:39.718050 [info] client : HumanClientApp.cpp:1545 : Sending OS request to browse directory: start /b "C:\Users\Geoff\Desktop\FreeOrionSDK5\FreeOrionAlt\"`

**Comment by dbenage-cx** [comment]
Dropped the `/b` option in case it is causing issues, might see a command window open when clicking links.

**Comment by geoffthemedio** [comment]
It does make a command console window pop up, but also doesn't open the directory window.

**Comment by dbenage-cx** [comment]
Trying without quotes and using preferred form of appending directory separator.

**Comment by geoffthemedio** [comment]
Seems to work.
````
10:49:54.453199 [info] client : HumanClientApp.cpp:1547 : Sending OS request to browse directory: start C:\Users\Geoff\AppData\Roaming\FreeOrion\save\

10:50:00.657318 [info] client : HumanClientApp.cpp:1547 : Sending OS request to browse directory: start C:\Users\Geoff\AppData\Roaming\FreeOrion\save\asdfШит\
````

**Comment by geoffthemedio** [comment]
How do I now link to a path with spaces in it?
````
<browsepath "C:\Users\Geoff\AppData\Roaming\FreeOrion\Dir With Spaces"</browsepath>
<browsepath PATH_DATA_USER/Dir With Spaces>Path With Spaces</browsepath>
````
Cause just an error message to appear from the rich text parse: `Error parsing rich text tags: expected end tag:browsepath got end of string.`

**Comment by dbenage-cx** [comment]
Needed to verify it was quoting and not final backslash or other issue.
Escaping quotes for cmd looks like it might be `^"`, `""`, `/"` or unneeded depending on context and options.

**Comment by geoffthemedio** [comment]
AppVeyor failure reproduced locally.

**Comment by geoffthemedio** [comment]
in console window:
`'"start"' is not recognized as an internal or external command, operable program or batch file.`
in log:
`23:54:25.571047 [info] client : HumanClientApp.cpp:1547 : Sending OS request to browse directory: ""start" "C:\Users\Geoff\Desktop\FreeOrionSDK5\FreeOrionAlt\""`

**Comment by dbenage-cx** [comment]
Stumbled across http://www.windowsinspired.com/how-a-windows-programs-splits-its-command-line-into-individual-arguments/ which was insightful (not a short article).

Trying escaping the path separator and retaining the second double quotes, else removing both if that fails: `command = L"start \"" + target;`.

**Comment by geoffthemedio** [comment]
build fails

**Comment by geoffthemedio** [comment]
I just get `Error parsing rich text tags: expected end tag:browsepath got end of string.` for the article with test links with this version.

**Comment by geoffthemedio** [comment]
Opens a new console window instead of a directory browse window.

**Comment by geoffthemedio** [comment]
Still opens a console window.

`    command = L"start \"\" \"" + target + L"\\\\\"";` works.

Based on https://superuser.com/questions/511486/how-to-start-open-a-file-folder-that-contains-space-in-its-name-through-command

**Comment by dbenage-cx** [comment]
Ty for finding a resolution and continuing to test.
`Fixup`ed up the temporary commits and improved the comments.

Omitted the `/b` option (prevent new console window), so as to use your unaltered form.
